# Latent-Taxonomic-Signatures
To explain the sudden appearance of species and a paucity of evidence for constant-rate gradualism, theories such as punctuated equilibrium have introduced the concept that evolution has more than one gear. In this theory, rapidly evolving populations enter stasis once speciation is completed and the resultant species become adapted. Instead of using traditional protein sequence alignment based methods to examine models for the evolution of species, here the frequency of 3-peptide motifs shared between otherwise unrelated species proteomes was determined using Latent Semantic Analysis. A positive correlation between semantic similarity of proteomes and species taxonomy was discovered. This new proteome feature was named Latent Semantic Signatures. Thus, any arbitrarily selected set of proteins would reflect the overall proteome signature for a given species, which was not reflected in any other genome level biases. Even proteins regarded as taxonomically restricted (e.g. orphan) were found to display these signatures. Evidence is also provided that shared adaptations (e.g. virus-host specificity) that are not restricted to single species also display latent proteome signatures and, that these proteome signatures can be used for functional discrimination with high degree of accuracy (e.g. between venomous and non-venomous animals). Thus, the discovery of latent signatures at the level of a species proteome independent of genome bias strengthens support for the punctuated equilibrium hypothesis.

The python code herein consists of various scripts used to process species proteome data originating from NCBI "nr" database (https://www.ncbi.nlm.nih.gov/refseq/about/nonredundantproteins/). Numerals at the end of each script indicate the order in which the code should be executed. Letters "a", "b" or "c" at the end, also indicate sequential order of execution. This is just a "rough" pipeline and due to numerous dependancies on external libraries (standalone BLAST+, HMMER 3, Pfam, ETE3 Taxonomy, Gensim etc.), each script should be executed only after all dependancies listed in the beginning of the scripts have been installed.  

If there are some issues with code execution, please contact me by sending me an email on astar@pbf.hr with the title "LTS code issue" and state the script you have trouble excecuting. 

These are the main scripts used to generate data which is included in paper "Latent semantic analysis supports Molecular Drive as an evolutionary force that shapes the species". The code used to generate figures is not included, but there is a web application called "Matrix of Life" based on this newly discovered feature, which is available for all to use and test at will with their own data without the need to install anything: http://matrix.pbf.hr/
