# Latent-Semantic-Signatures
Matrix outlook on the Tree of Life
To systematically explore hypothesis of species evolution through cohesive mode of molecular drive we constructed over 50,000 species models based on Latent Semantic Analysis. We substituted sequence alignment with natural language processing and distributional semantics in order to test for existence of cohesive information between sequences that cannot be aligned, but share taxonomy as common denominator. 

This revealed distributional similarity between gene translations belonging to species proteomes. Despite being members of different gene families, all translated sequences exhibited shared information content related to their shared proteome taxonomic lineage in form of a 3-peptide motif frequency signature. We present evidence of large-scale molecular drive contribution not limited to multigene families and other paralogous sequences, highlighting its importance as a full-fledged evolutionary force.

The python code herein consists of various scripts used to process species proteome data originating from NCBI "nr" database (https://www.ncbi.nlm.nih.gov/refseq/about/nonredundantproteins/). Numerals at the end of each script indicate the order in which the code should be executed. Letters "a", "b" or "c" at the end, also indicate sequential order of execution. This is just a "rough" pipeline and due to numerous dependancies on external libraries (standalone BLAST+, HMMER 3, Pfam, ETE3 Taxonomy, Gensim etc.), each script should be executed only after all dependancies listed in the beginning of the scripts have been installed.  

If there are some issues with code execution, please contact me by sending me an email on astar@pbf.hr with the title "LTS code issue" and state the script you have trouble excecuting. 
